presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -689,26 +710,59 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
         Object left = ExpressionInterpreter.expressionOptimizer(comparison.getLeft(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
         Object right = ExpressionInterpreter.expressionOptimizer(comparison.getRight(), metadata, session, expressionTypes).optimize(NoOpSymbolResolver.INSTANCE);
 
-        if (left instanceof SymbolReference && !(right instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) left, comparison.getType(), new NullableValue(expressionTypes.get(comparison.getRight()), right)));
+        if (left instanceof Expression && right instanceof Expression) {
+            return Optional.empty();
+        }
+
+        if (!(left instanceof Expression) && !(right instanceof Expression)) {
+            return Optional.empty();
+        }
+
+        if (left instanceof Expression) {
+            return toNormalizedSimpleComparison(metadata, expressionTypes, comparison.getLeft(), comparison.getType(), new NullableValue(expressionTypes.get(comparison.getRight()), right));
+        }
+        else {
+            return toNormalizedSimpleComparison(metadata, expressionTypes, comparison.getRight(), comparison.getType().flip(), new NullableValue(expressionTypes.get(comparison.getLeft()), left));
         }
-        if (right instanceof SymbolReference && !(left instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) right, comparison.getType().flip(), new NullableValue(expressionTypes.get(comparison.getLeft()), left)));
+    }
+
+    private static Optional<NormalizedSimpleComparison> toNormalizedSimpleComparison(Metadata metadata, IdentityHashMap<Expression, Type> expressionTypes, Expression nonLiteral, ComparisonExpressionType comparisonType, NullableValue literal)
+    {
+        boolean nonCoercionCastRemoved = false;
+        while (true) {
+            if (nonLiteral instanceof SymbolReference) {
+                return Optional.of(new NormalizedSimpleComparison((SymbolReference) nonLiteral, comparisonType, literal, nonCoercionCastRemoved));
+            }
+            else if (nonLiteral instanceof Cast) {
+                if (!isCoercion(metadata, expressionTypes, (Cast) nonLiteral)) {
```

Comment:
@erichwang  I looks like you are right here.

Also it seems that for now we should skip this optmisiation alltogether for cases where CAST on symbol is not coercion.
This code is buggy for cases like:

`CAST(decimal_23_3_column AS BIGINT) = BIGINT '23'`

No it would enter saturated floor cast path.
Which would lead to rewriting expression to:

`decimal_23_3_column = DECIMAL '23.000'`

which is clearly wrong.

The expected tuple domain should be `[23.000, 23.999]`

We will work on fixing this. And later we can extend the mechanism for handling non-coercions correctly.

cc: @KBP-TDC, @arhimondr